sherpa/utils/random.py

```diff
@@ -27,17 +27,44 @@
 
 """
 
+from typing import Optional, Sequence, SupportsFloat, Union, \
+    overload
+
 import numpy as np
 
 from .numeric_types import SherpaFloat
 
+# This should probably be an explicit type alias but for now leave it
+# like this. Users are expected to use the Generator rather than
+# RandomState form, but both are kept in (e.g. for testing).
+#
+# Once Python 3.10 or 3.12 is the minimum we can make this an actual
+# alias that we can export. For now this "alias" can be used by other
+# modules but the documentation is not ideal and the name may change.
+#
+RandomType = Union[np.random.Generator, np.random.RandomState]
+
 
 __all__ = ("chisquare", "choice", "integers",
            "multivariate_normal", "normal", "poisson_noise",
            "random", "standard_normal", "uniform")
 
 
-def poisson_noise(x, rng=None):
+@overload
+def poisson_noise(x: SupportsFloat,
+                  rng: Optional[RandomType] = None
+                  ) -> SherpaFloat:
+    ...
+
+@overload
+def poisson_noise(x: Sequence[SupportsFloat],
+                  rng: Optional[RandomType] = None
+                  ) -> np.ndarray:
+    ...
+
+def poisson_noise(x: Union[SupportsFloat, Sequence[SupportsFloat]],
```

Comment:
I had thought I'd seen stuff where the "actual" function still had typing arguments, but I'm happy to be proven wrong.

Looking into https://peps.python.org/pep-0484/ it does say

> The @overload-decorated definitions are for the benefit of the type checker only, since they will be overwritten by the non-@overload-decorated definition, while the latter is used at runtime but should be ignored by a type checker.

which would match your description (and makes sense).

Reply:
So, I've removed this relpication, which is nice. As part of this I've extended the use of overload as some of the functions would return a scalar if `size=None` and an ndarray otherwise, so we want to encode this behavior which `overload` does.